proton/cert_pinning.py

```diff
@@ -0,0 +1,171 @@
+    """HTTPSConnectionPool that verifies the certificate for each connection"""
+    def __init__(
+        self,
+        host,
+        hash_dict,
+        port=None,
+        strict=False,
+        timeout=Timeout.DEFAULT_TIMEOUT,
+        maxsize=1,
+        block=False,
+        headers=None,
+        retries=None,
+        _proxy=None,
+        _proxy_headers=None,
+        key_file=None,
+        cert_file=None,
+        cert_reqs=None,
+        key_password=None,
+        ca_certs=None,
+        ssl_version=None,
+        assert_hostname=None,
+        assert_fingerprint=None,
+        ca_cert_dir=None,
+        **conn_kw
+    ):
+        self.hash_dict = hash_dict
+        super(TLSPinningHTTPSConnectionPool, self).__init__(
+            host,
+            port,
+            strict,
+            timeout,
+            maxsize,
+            block,
+            headers,
+            retries,
+            _proxy,
+            _proxy_headers,
+            key_file,
+            cert_file,
+            cert_reqs,
+            key_password,
+            ca_certs,
+            ssl_version,
+            assert_hostname,
+            assert_fingerprint,
+            ca_cert_dir,
+            **conn_kw
+        )
+
+    def _validate_conn(self, conn):
+        r = super(TLSPinningHTTPSConnectionPool, self)._validate_conn(conn)
+
+        sock = conn.sock
+
+        pem_certificate = self.get_certificate(sock)
+
+        if self.is_session_secure(pem_certificate, conn, self.hash_dict):
+            return r
+
+    def is_session_secure(self, cert, conn, hash_dict):
+        """Checks if connection is secure"""
+
+        cert_hash = self.extract_hash(cert)
+
+        if not self.validate_hash(cert_hash, hash_dict):
+            # Also generate a report
+            conn.close()
+            raise TLSPinningError("Insecure connection")
+
+        return True
+
+    def validate_hash(self, cert_hash, hash_dict):
+        """Validates the hash agains a known list of hashes/pins"""
+        try:
+            hash_dict[self.host].index(cert_hash)
+        except (ValueError, KeyError, TypeError):
+            return False
+        else:
+            return True
+
+    def get_certificate(self, sock):
+        """Extracts and converts certificate to PEM"""
+        certificate_binary_form = sock.getpeercert(True)
+        pem_certificate = DER_cert_to_PEM_cert(certificate_binary_form)
+
+        return pem_certificate
+
+    def extract_hash(self, cert):
+        """Extracts the encrypted hash from the certificate"""
+        cert_obj = crypto.load_certificate(crypto.FILETYPE_PEM, cert)
+
+        pubkey_obj = cert_obj.get_pubkey()
+
+        pubkey = crypto.dump_publickey(crypto.FILETYPE_ASN1, pubkey_obj)
+        spki_hash = hashlib.sha256(pubkey).digest()
+        cert_hash = base64.b64encode(spki_hash).decode('utf-8')
+
+        return cert_hash
+
+
+class TLSPinningPoolManager(PoolManager):
+    """
+    Custom PoolManager that attaches
```

Comment:
`docstring` are usually formatted as
* first line: 80 chars with imperative verb
* new line
* longer description (with parameters' specification, return values, etc)

or just the first line.